### ISSUE TITLE
bug with serializing filter

### ISSUE THREAD

**Issue by yurks** [ethanresnick/json-api-custom-query#8]
```js
serializeFilter(parseFilter("(:exists,fld1)"))
//-> "(:exists,fld1)"

serializeFilter(parseFilter("(:exists,fld1,fld2)"))
//-> "(fld1,:exists,fld2)"  <-- wrong

serializeFilter(parseFilter("(:exists,fld1,fld2,fld3)"))
//-> "(:exists,fld1,fld2,fld3)"


```

**Comment by ethanresnick** [comment]
This is actually not a bug, exactly. 

The syntax [allows](https://github.com/ethanresnick/json-api-custom-query/blob/master/src/parsing/grammar.pegjs#L40) for leading (i.e., prefixed/polish notation) operators or, when there are precisely two arguments, it allows the operator to be infixed. Regardless of whether the operator is infixed or not, the parse result is the same (so `(:exists,fld1,fld2)` is treated exactly like `(fld1,:exists,fld2)`).

The infixed notation grew out of an earlier form where all of the operators were binary (which most of them still are), and I thought is was convenient for readability to allow forms like `(field1,:eq,field2)`.

Now, the issue here really is that the serializer doesn't know that your `:exists` operator takes an arbitrary number of arguments. So, when it sees two arguments, it assumes it's a binary operator and serializes it to use the "normal form" for these binary expressions, which is the infix version. 

The only fix for this would be to pass the serializer info about the arity of each of your operators. Personally, I'm not in favor of making that information _required_, as I think it's really important that parsing and serializing can currently happen in a valid way without any extra information being provided that isn't in the syntax itself. However, I would support an optional argument to the serializer that describes the arity of each operator so that, for non-binary operators, it can serialize `(:exists,fld1,fld2)`. That would be purely for aesthetic reasons, but aesthetics count too.

**Comment by yurks** [comment]
yes, argument for serializer for non-binary operators would be a good options. Or, maybe better, something like .serializeNonBinaryFilter(), as I resolve it for myself. Couldn't pull a patch for now, 'casue not familiar with typescript and make it in built files for my needs.
